cohorts/cohort.py

```diff
@@ -470,23 +489,117 @@ def load_variants(self, patients=None, filter_fn=None):
             Dictionary of patient_id to VariantCollection
         """
         filter_fn = first_not_none_param([filter_fn, self.filter_fn], no_filter)
+        if filter_fn is None:
+            filter_fn_name = 'None'
+        else:
+            filter_fn_name = filter_fn.__name__
+        logger.debug('loading variants with filter_fn: {}'.format(filter_fn_name))
         patient_variants = {}
 
         for patient in self.iter_patients(patients):
-            variants = self._load_single_patient_variants(patient, filter_fn)
+            variants = self._load_single_patient_variants(patient, filter_fn, **kwargs)
             if variants is not None:
                 patient_variants[patient.id] = variants
         return patient_variants
+    
+    
+    def _hash_filter_fn(self, filter_fn, **kwargs):
+        if filter_fn is None:
+            return 'filter-none'
+        filter_fn_name = filter_fn.__name__
+        logger.debug('Computing hash for filter_fn: {} with kwargs {}'.format(filter_fn_name, str(dict(**kwargs))))
+        # function source code
+        fn_source = str(dill.source.getsource(filter_fn))
+        pickled_fn_source = pickle.dumps(fn_source)
+        # kwarg values
+        kw_dict = dict(**kwargs)
+        kw_hash = list()
+        if not kw_dict:
+            kw_hash = ['default']
+        else:
+            [kw_hash.append('{}-{}'.format(key, h)) for (key, h) in sorted(kw_dict.items())]
+        # closure vars
+        closure = 'null'
+        nonlocals = inspect.getclosurevars(filter_fn).nonlocals
+        for (key, val) in nonlocals.items():
+            if inspect.isfunction(val):
+                closure = '{}-{}'.format(self._hash_filter_fn(val).__name__, self._hash_filter_fn(val))
+        # construct final hashed_fn
+        hashed_fn = '.'.join(['-'.join([filter_fn.__name__,
+                                        int(hashlib.sha1(pickled_fn_source).hexdigest(), 16) % (10 ** 11)]),
+                              '.'.join(kw_hash),
+                              closure]
+                            )
+        return hashed_fn
+            
+        
+    
+    def _load_single_patient_variants(self, patient, filter_fn, use_cache=True, **kwargs):
```

Comment:
Add comment/docstrings stating that this function deals with loading/saving to the filter_fn cache, while the merged variant function has the unfiltered variant cache?